app/Platform/Controllers/DeveloperRecentActivityController.php

```diff
@@ -0,0 +1,283 @@
+                }
+            );
+    }
+
+    private function buildFriendlyTimestampLabel(?Carbon $unlockedAt, ?Carbon $unlockedHardcoreAt): string
+    {
+        // Use the hardcore unlock date if available, otherwise use the normal unlock date.
+        $timestamp = $unlockedHardcoreAt ?? $unlockedAt;
+
+        // If there is no timestamp, return an empty string.
+        if (!$timestamp) {
+            return '';
+        }
+
+        // Compare the date with now.
+        if ($timestamp->isToday()) {
+            // If the timestamp is from today, use the 'timeago' format.
+            return $timestamp->diffForHumans();
+        } elseif ($timestamp->isYesterday()) {
+            // If the timestamp is from yesterday, use the 'Yesterday at' format.
+            return 'Yesterday at ' . $timestamp->format('g:ia');
+        } else {
+            // If the timestamp is older, use the 'F j Y, g:ia' format.
+            return $timestamp->format('F j Y, g:ia');
+        }
+    }
+
+    private function buildLeaderboardEntriesForDevBaseQuery(User $targetUser): mixed
+    {
+        return LeaderboardEntryLegacy::query()
+            ->join('LeaderboardDef', 'LeaderboardDef.ID', '=', 'LeaderboardEntry.LeaderboardID')
+            ->where('LeaderboardDef.Author', $targetUser->User);
+    }
+
+    private function fetchAwardsContributedForDev(mixed $allUserGameIds): int
+    {
+        return $this->buildAwardsForDevBaseQuery($allUserGameIds)->count();
+    }
+
+    private function fetchLeaderboardEntriesContributedForDev(User $targetUser): int
+    {
+        return $this->buildLeaderboardEntriesForDevBaseQuery($targetUser)->count();
+    }
+
+    private function fetchRecentAwardsForDev(mixed $allUserGameIds, int $limit = 200): mixed
+    {
+        $mostRecentAwards = $this->buildAwardsForDevBaseQuery($allUserGameIds)
+            ->orderByDesc('pb.AwardDate')
+            ->take($limit)
+            ->get();
+
+        return $this->attachRecentAwardRowsMetadata($mostRecentAwards);
+    }
+
+    private function fetchRecentLeaderboardEntriesForDev(User $targetUser, int $limit = 200): mixed
+    {
+        $mostRecentLeaderboardEntries = $this->buildLeaderboardEntriesForDevBaseQuery($targetUser)
+            ->orderByDesc('LeaderboardEntry.DateSubmitted')
+            ->take($limit)
+            ->get([
+                'LeaderboardDef.ID', 'LeaderboardDef.GameID', 'LeaderboardDef.Format', 'LeaderboardDef.Title', 'LeaderboardDef.Description',
+                'LeaderboardEntry.LeaderboardID', 'LeaderboardEntry.UserID', 'LeaderboardEntry.Score', 'LeaderboardEntry.DateSubmitted',
+            ]);
+
+        return $this->attachRecentLeaderboardEntryRowsMetadata($mostRecentLeaderboardEntries);
+    }
+
+    private function fetchRecentUnlocksForDev(mixed $allUserAchievementIds, int $limit = 200): mixed
+    {
+        $mostRecentUnlocks = PlayerAchievement::whereIn('achievement_id', $allUserAchievementIds)
```

Comment:
This `whereIn()` can result in a very large query to the database (in terms of raw content size), however I couldn't come up with a way to do this that was more performant.

Reply:
On my machine:
```
> select pa.* from player_achievements pa inner join Achievements ach
on ach.ID=pa.achievement_id where ach.Author='Jamiras' order by pa.unlocked_at desc limit 200;
...
200 rows in set (0.158 sec)
```

vs.
```
select * from `player_achievements` where `achievement_id` in (50421, ..., 363389) order by `unlocked_at` desc limit 200
...
200 rows in set (0.164 sec)
```

The difference is negligible, and the join does appear to win out.

Reply:
I think something is just up with my local DB. The first time I run this query, it takes an eternity:

![Screenshot 2023-11-10 at 5 38 18 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/c772750c-b700-42a2-a09a-7beb80308f93)

However, subsequent runs are much quicker:

![Screenshot 2023-11-10 at 5 38 30 PM](https://github.com/RetroAchievements/RAWeb/assets/3984985/db7abd54-7ee6-4121-a382-a8dc8189db0b)

Reply:
I see that behavior too if I run the query directly against other random devs. Not sure what's going on. Probably safer to just go with the achievement list.

It doesn't seem to matter, but you should only be selecting `pa.*`.  Selecting `*` also pulls in all the information from the achievements (including the definition which might be quite large).

Reply:
Got it. Indeed, selecting `pa.*` made no difference for that first hit. When I revert to using `whereIn()`, I can't repro the slow first hit anymore.